### ISSUE TITLE
Duplicate names overriding eachother

### ISSUE THREAD

**Issue by Zyles** [thephpleague/container#117]
I created two ServiceProviders with their own namespace.

But if I register the same key (hellocommands) in both (notice HelloWorld and HelloWorld2).

This code exists in their respective ServiceProvider.php class, I just type both here to illustrate.

```
$this->getContainer()->add('hellocommands', 'Bot\HelloWorld\HelloCommands');
$this->getContainer()->add('hellocommands', 'Bot\HelloWorld2\HelloCommands');
```

Only the latter exists in the container.

When I do this:

```
$this->getContainer()->add('1_hellocommands', 'Bot\HelloWorld\HelloCommands');
$this->getContainer()->add('2_hellocommands', 'Bot\HelloWorld2\HelloCommands');
```

Both shows up.

Is this normal?

How would I go about forcing modules of my application to have unique key names to avoid this since it seems it does not use namespace to separate them?

**Comment by philipobenito** [comment]
The container retrieves items based on key name, if you add a key twice, one will override the other. It is down to your application to handle your key names.

**Comment by Zyles** [comment]
Well, is it possible to get a list of registered key names then so I can throw an exception?

**Comment by philipobenito** [comment]
No that's not something you can do, keys are stored in several different ways and  are also skewed if any delegate containers are enabled so it's impossible to give an accurate representation. 

The container is aimed at storing and wiring together objects and services, it doesn't sound like this is what you're using it for?

**Comment by Zyles** [comment]
I believe I am.

My architecture looks like this:

The main bot is running very bare bones with a DI and some basic functionality, and there are modules to extend the bot. Think composer packages.

When you install a module you register the ServiceProvider with the main $bot object. This serviceprovider can typically host a number of different classes such as commands or services.

Then you will be able to access the modules through $bot->mymodule->command()

But what if two packages happen to register two keys with the same names? Then it will break.

I guess I can create a function to look at the DI object or something before running addServiceProvider()

**Comment by philipobenito** [comment]
You can run the `has` method to check a specific key, you just can't get a full list.

**Comment by Zyles** [comment]
I had to drop the whole key idea, couldn't get a good design going. Since I ran into catch 22 because I could not get the key names.

Is there at least a function that can check if a serviceprovider is already registered?

Couldn't find any. And I can not access $this->bot->providers->registered since it is protected...

Or maybe I can just register the same provider multiple times, since it seems it does it for every class. Maybe it doesn't matter.
